hera_cal/redcal.py

```diff
@@ -269,6 +337,127 @@ def multiply_by_gains(target, gains, target_type='vis'):
+                conv_crit: the convergence criterion evaluated at the final iteration
+            sol: a dictionary of complex solutions with variables as keys
+        """
+        sol = self.sol0
+        terms = [(linsolve.get_name(gi),linsolve.get_name(gj),linsolve.get_name(uij)) 
+            for term in self.all_terms for (gi,gj,uij) in term]
+        dmdl_u = self._get_ans0(sol)
+        chisq = sum([np.abs(self.data[k]-dmdl_u[k])**2 * self.wgts[k] for k in self.keys])
+        # variables with '_u' are flattened and only include pixels that need updating
+        dmdl_u = {k:v.flatten() for k,v in dmdl_u.items()}
+        # wgts_u hold the wgts the user provides.  dwgts_u is what is actually used to wgt the data
+        wgts_u = {k: (v * np.ones(chisq.shape, dtype=np.float32)).flatten() for k,v in self.wgts.items()}
+        sol_u = {k:v.flatten() for k,v in sol.items()}
+        iters = np.zeros(chisq.shape, dtype=np.int)
+        conv = np.ones_like(chisq)
+        update = np.where(chisq > 0)
+        for i in range(1,maxiter+1):
+            if verbose: print('Beginning iteration %d/%d' % (i,maxiter))
+            if (i % check_every) == 1:
+                # compute data wgts: dwgts = sum(V_mdl^2 / n^2) = sum(V_mdl^2 * wgts)
+                # don't need to update data weighting with every iteration
+                dwgts_u = {k: dmdl_u[k] * dmdl_u[k].conj() * wgts_u[k] for k in self.keys}
+                sol_wgt_u = {k:0 for k in sol.keys()}
+                for k,(gi,gj,uij) in zip(self.keys, terms):
+                    w = dwgts_u[k]
+                    sol_wgt_u[gi] += w
+                    sol_wgt_u[gj] += w
+                    sol_wgt_u[uij] += w
+                dw_u = {k:v[update] * dwgts_u[k] for k,v in self.data.items()}
+            sol_sum_u = {k:0 for k in sol_u.keys()}
+            for k,(gi,gj,uij) in zip(self.keys, terms):
+                # compute sum(wgts * V_meas / V_mdl)
+                numerator = dw_u[k] / dmdl_u[k]
+                sol_sum_u[gi] += numerator
+                sol_sum_u[gj] += numerator.conj()
+                sol_sum_u[uij] += numerator
+            new_sol_u = {k: v * ((1 - self.gain) + self.gain * sol_sum_u[k]/sol_wgt_u[k]) 
+                            for k,v in sol_u.items()}
+            dmdl_u = self._get_ans0(new_sol_u)
+            # check if i % check_every is 0, which is purposely one less than the '1' up at the top of the loop
+            if i < maxiter and (i < check_after or (i % check_every) != 0):
+                # Fast branch when we aren't expensively computing convergence/chisq
+                sol_u = new_sol_u
+            else:
+                # Slow branch when we compute convergence/chisq
+                new_chisq_u = sum([np.abs(v[update]-dmdl_u[k])**2 * wgts_u[k] for k,v in self.data.items()])
+                chisq_u = chisq[update]
+                gotbetter_u = (chisq_u > new_chisq_u)
```

Comment:
This block with it's variety of similarly named variables (`gotbetter_u`, `where_gotbetter_u`, and `update_where`) could use a comment or two explaining what's going on.